### PR TITLE
RUBY-3010 Allow unlimited connection pool

### PR DESCRIPTION
**Pull request by comandeo**
https://jira.mongodb.org/browse/RUBY-3010
https://jira.mongodb.org/browse/RUBY-3011

### CODE REVIEW

File: docs/reference/create-client.txt [hunk 1]
```diff
@@ -548,8 +548,9 @@ Ruby Options
 
    * - ``:max_pool_size``
      - The maximum size of the connection pool for each server.
+       Setting this option to zero sets no limit for a maximum size of the connection pool.
```
Comment (p-mongo) [inline docs/reference/create-client.txt:551]:
```suggestion
       Setting this option to zero removes the max size limit from the connection pool, permitting it to grow to any number of connections.
```

File: spec/mongo/server/connection_pool_spec.rb [hunk 2]
```diff
@@ -83,11 +83,21 @@
...
+      end
+    end
+
+    context 'when min size is provided and max size is zero (unlimited)' do
+      let (:options) do
+        { min_size: 10, max_size: 0 }
+      end
+
+      it 'sets max size to zero (unlimited)' do
+        expect(pool.max_size).to eq(0)
```
Comment (p-mongo) [inline spec/mongo/server/connection_pool_spec.rb:100]: There should also be an assertion that the min size is respected and unchanged from user-provided value, in this or a new test.

File: lib/mongo/server/connection_pool.rb [hunk 3]
```diff
@@ -377,7 +378,7 @@ def check_out(connection_global_id: nil)
                 #
                 # Ruby does not allow a thread to lock a mutex which it already
                 # holds.
-                if unsynchronized_size < max_size
+                if max_size == 0 || unsynchronized_size < max_size
```
Comment (p-mongo) [inline lib/mongo/server/connection_pool.rb:381]: I am not a fan of max_size attribute of the pool being zero but the alternative I can think of (have max_size be nil for an unlimited pool) is also confusing: nil max_size in constructor argument is changed to max_size=20, max_size=0 in constructor argument is changed to max_size nil. Unless you have some other ideas we can leave this the way it is typed here for now.

**Review by p-mongo — APPROVED** [review]
There are 2 test failures in ./spec/mongo/client_construction_spec.rb which need to be remedied.

It wouldn't hurt to have some more tests giving 0 as max size to client constructor and verifying either that the zero is correctly forwarded to pool (via method expectations) or asserting actual pool attributes.

File: lib/mongo/server/connection_pool.rb (hunk 3 above)
Comment (comandeo) [reply lib/mongo/server/connection_pool.rb:381]: We could theoretically use `Float::INFINITY` for the `max_size`, but it is (obviously) float.

**Review by Neilshweky — APPROVED** [review]
